pkg/configurator/validating_webhook.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	// Decode the configmap from the request
+	var configMap corev1.ConfigMap
+	if _, _, err := deserializer.Decode(req.Object.Raw, nil, &configMap); err != nil {
+		log.Error().Err(err).Msgf("Error unmarshaling request to configmap in namespace %s", req.Namespace)
+		return webhook.AdmissionError(err)
+	}
+
+	log.Trace().Msgf("Validation request: (new object: %v) (old object: %v)", string(req.Object.Raw), string(req.OldObject.Raw))
+
+	//default response
+	resp := &v1beta1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{Reason: ""},
+		UID:     req.UID,
+	}
+
+	//Allow if not a change to osm-config
+	if req.Name != constants.OSMConfigMap {
+		return resp
+	}
+
+	whc.validate(configMap, resp)
+
+	return resp
+}
+
+//validate performs all the checks to the configmap fields and rejects as necessary
+func (whc *webhookConfig) validate(configMap corev1.ConfigMap, resp *v1beta1.AdmissionResponse) *v1beta1.AdmissionResponse {
+	for field, value := range configMap.Data {
+		if !checkBoolFields(field, value, boolFieldsInConfigMap) {
+			reasonForDenial(resp, MustBeBool, field)
+		}
+		if field == "envoy_log_level" && !checkEnvoyLogLevels(field, value, ValidEnvoyLogLevels) {
+			reasonForDenial(resp, MustBeValidLogLvl, field)
+		}
+		if field == "service_cert_validity_duration" {
+			t, err := time.ParseDuration(value)
+			if err != nil {
+				reasonForDenial(resp, MustBeValidTime, field)
+			}
+			if t.Hours() > hrInAYear {
+				reasonForDenial(resp, MustBeLessThanAYear, field)
+			}
+		}
+		if field == "tracing_address" && !matchAddrSyntax(field, value) {
+			reasonForDenial(resp, MustFollowSyntax, field)
+		}
+		if field == "tracing_port" {
+			portNum, err := strconv.Atoi(value)
+			if err != nil {
+				reasonForDenial(resp, MustbeInt, field)
+			}
+			if portNum < 0 || portNum > maxPortNum {
+				reasonForDenial(resp, MustBeInPortRange, field)
+			}
+		}
+	}
+
+	originalConfigMap, _ := whc.kubeClient.CoreV1().ConfigMaps(whc.osmNamespace).Get(context.TODO(), constants.OSMConfigMap, metav1.GetOptions{})
+
+	for metadataAnnotation, val := range configMap.ObjectMeta.Annotations {
+		if originalConfigMap.Annotations[metadataAnnotation] != val {
+			reasonForDenial(resp, CannotChangeMetadata, metadataAnnotation)
+		}
+	}
+	for metadataLabels, val := range configMap.ObjectMeta.Labels {
+		if originalConfigMap.Labels[metadataLabels] != val {
+			reasonForDenial(resp, CannotChangeMetadata, metadataLabels)
+		}
+	}
+	return resp
+}
+
+//matchAddrSyntax checks that string follows hostname syntax
+func matchAddrSyntax(configMapField, configMapValue string) bool {
+	syntax := regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
+
+	return syntax.Match([]byte(configMapValue))
+}
+
+//checkEnvoyLogLevels checks that the field value is a valid log level
+func checkEnvoyLogLevels(configMapField, configMapValue string, levels []string) bool {
+	valid := false
```

Comment:
Its better if we avoid passing the valid log levels as an argument here since it is going to be a static list at all times. Use `validEnvoyLogLevels` directly within this function.